src/Symfony/Component/Dotenv/Dotenv.php

```diff
@@ -35,15 +35,47 @@ final class Dotenv
     private $data;
     private $end;
     private $values;
-    private $usePutenv;
+    private $envKey;
+    private $debugKey;
+    private $prodEnvs = ['prod'];
```

Comment:
perhaps for consistency, define `$testEnvs = ['test'] + $defaultEnv = 'dev';` as well? Then cleanup the arg list in `loadEnv()`

Reply:
I considered that, but I think it's not worth the upgrade trouble.

Reply:
it would avoid a half stateful / stateless API, and perhaps is worth it from the component's design POV.

Reply:
do we prefer mutability? or what about `new Dotenv('APP_ENV', 'APP_DEBUG', 'dev', ['prod'], ['test'])` :/

Reply:
We have no issues with mutability when it's purpose isto configure the instance.
Using many constructor args is a pain when you only want to change the last one.
Setters are better here IMHO.

Reply:
i undertand the technical aspect, im curious if a user should change the context between loading envs (vs using a dedicated instance per context). In practice the context is a fixed concern.

Reply:
In practice, people will figure out what they need :)
I considered alternatives before deciding for this design: it looks the best compromise to me.

Reply:
IMHO it's component design first :) i strongly feel the _right_ design is

```php
$dotenv = new Dotenv('APP_ENV', 'APP_DEBUG', 'dev', ['prod'], ['test']);
$dotenv->load('.env', $putenv = true);
$dotenv->loadAll(['.env'], $putenv = true);
```

but im fine either way, my only use case is a generated one :)

Reply:
I understand, that's what I meant by "compromise". We're not designing from the ground up, and in practice this is the most seamless.